### PR TITLE
Adding retry and retry back off to allow retry in case executor fails

### LINKED ISSUE #19: Delay in vm configuration leads ssh to fail
## Actual behavior
```
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] Executing command [sh -c ". ./scripts/minikube-minishift-setup-env.sh minikube && . ./scripts/minikube-minishift-all-tests.sh minikube"] 
INFO[2021-04-08T13:04:00Z]                                              
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32]                              
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] .--------------------------------------- 
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] | Building odo binaries      
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] '--------------------------------------- 
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32]                              
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] ++ make bin                  
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] Command Error failed to wait for command completion failed to wait ssh command: <nil>, failing gracefully 
INFO[2021-04-08T13:04:00Z] [ssh:Fedora 32] !!!run failed, skipping!!!   
INFO[2021-04-08T13:04:00Z]                                              
INFO[2021-04-08T13:04:00Z] 2021/04/08 13:03:50 : failed due to err Failed the test, see logs above ^ 
INFO[2021-04-08T13:04:00Z] {"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-04-08T13:03:50Z"} 
INFO[2021-04-08T13:04:00Z] error: failed to execute wrapped command: exit status 1 
```

For now we are adding sleep just after vm creation https://gitlab.cee.redhat.com/openshiftdo/psi-jenkins/-/blob/master/scripts/configure-openstack-vm.sh#L33 , so that VM gets fully set to login. However its not helping much to fix the flake.

## Suggestion
Setup a retry loop for sshing and once the VM are fully configured it will do ssh into the machine, should fix the issue.

## Any logs, error output, etc?
More details: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_odo/4561/pull-ci-openshift-odo-master-psi-kubernetes-integration-e2e/1380142723025080320#1:build-log.txt%3A70

### PR DESCRIPTION
**Pull request by mohammedzee1000**
Note this is only restricted to executor fails, not command fails
Fix #19 
Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

### CODE REVIEW

File: pkg/ci-firewall/cli/worker/work.go [hunk 1]
```diff
@@ -138,6 +141,9 @@ func (wo *WorkOptions) Validate() (err error) {
 	if (wo.gitUser != "" && wo.gitEmail == "") || (wo.gitEmail != "" && wo.gitUser == "") {
 		return fmt.Errorf("both git user and git email must be provided together or neither of them")
 	}
+	if wo.retryLoopCount < 1 {
+		return fmt.Errorf("retry cound should be a natural number i.e >= 1")
```
Comment (prietyc123) [inline pkg/ci-firewall/cli/worker/work.go]: Typo `cound` should be `count`

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/ci-firewall/cli/worker/work.go [hunk 2]
```diff
@@ -197,6 +203,8 @@ func NewWorkCmd(name, fullname string) *cobra.Command {
 	cmd.Flags().BoolVar(&o.redact, "redact", true, "if true, then injected envs and ip addresses are redacted from logs sent over queue. Default is true")
 	cmd.Flags().StringVar(&o.gitUser, "gituser", "", "The git user you want to configure for repo")
 	cmd.Flags().StringVar(&o.gitEmail, "gitemail", "", "The email of git user you want to configure on repo")
+	cmd.Flags().IntVar(&o.retryLoopCount, "retryloopcount", 3, "The number of times we retry command if it fails. Defaults to 3")
+	cmd.Flags().DurationVar(&o.retryLoopBackoff, "retryloopbackoff", 10 * time.Second, "The base amount of time to back off before retry on command execution due to execution errors. Defaults to 10 seconds. Actual backoff will be currentcount * backoff duration")
```
Comment (prietyc123) [inline pkg/ci-firewall/cli/worker/work.go:207]: You mean 3rd attempt will start after 30 sec of the second execution failure

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/worker/worker.go [hunk 3]
```diff
@@ -179,50 +191,70 @@ func (w *Worker) printAndStreamCommand(tags []string, cmdArgs []string) error {
...
+			rdr, err := ex.InitCommand(workDir, cmd, util.EnvMapCopy(w.envVars), w.tags)
+			if err != nil {
+				errList = append(errList, fmt.Errorf("failed to initialize executor %w", err))
+				continue
+			}
+			defer ex.Close()
+			done := make(chan error)
+			go func(done chan error) {
+				for {
+					data, err := rdr.ReadString('\n')
```
Comment (prietyc123) [inline pkg/worker/worker.go:229]: Can we add more comments.

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/worker/worker.go (hunk 3 above)
Comment (mohammedzee1000) [reply pkg/worker/worker.go:229]: ok

File: pkg/ci-firewall/cli/worker/work.go (hunk 2 above)
Comment (mohammedzee1000) [reply pkg/ci-firewall/cli/worker/work.go:207]: yes

File: pkg/ci-firewall/cli/worker/work.go (hunk 1 above)
Comment (mohammedzee1000) [reply pkg/ci-firewall/cli/worker/work.go]: fixing will push in a minute

File: pkg/ci-firewall/cli/worker/work.go (hunk 2 above)
Comment (mohammedzee1000) [reply pkg/ci-firewall/cli/worker/work.go:207]: You can reduce it too if you want by maybe overriding default value with 5s so 2nd attempt will be after 10s and 3rd after 15s for eg

File: pkg/worker/worker.go (hunk 3 above)
Comment (mohammedzee1000) [reply pkg/worker/worker.go:229]: done

File: pkg/ci-firewall/cli/worker/work.go (hunk 2 above)
Comment (prietyc123) [reply pkg/ci-firewall/cli/worker/work.go:207]: Make sense

**Comment by prietyc123** [comment]
Looks good to me

**Comment by anandrkskd** [comment]
Looks good to me

**Pull request merged**